### DISCUSSION TITLE
How does Koa determine if cookies are going over a secure connection?

### DISCUSSION THREAD

**Discussion by christianjuth** [koajs/koa#1826] [Q&A]
I'm deploying my Koa server to Fly.io. My domain is https, but I think whatever proxy Fly.io has set up is stripping away the encryption layer – I don't have much experience with proxies.

As far as I can tell, Fly.io [is doing everything correctly](https://fly.io/docs/networking/request-headers/#x-forwarded-proto). 

I'm also setting  `proxy.true` when creating my fly server.

```javascript
const app = new Koa({
  proxy: true,
})
```

I have the following options when I set my cookie.

```javascript
const option = ctx => ({
  secure: ctx.secure, // Set secure based on the request's secure property
  sameSite: ctx.secure ? 'none' : 'strict', // Set SameSite to 'none' for secure requests
})
app.use((ctx, next) => session(getSessionConfig(ctx), app)(ctx, next)) // add session cookie store

```

However, cookies package continues to reject the secure cookie saying `cannot send secure cookie over unencrypted connection`.

But I found this workaround:

```javascript
// THIS IS A HACK
app.use((ctx, next) => {
  ctx.cookies.secure = ctx.secure
  return next()
})
```

First, I'm wondering if there are any obvious security issues with this. Does `ctx.secure = true` imply that it's ok to send a secure cookie? Note, my fly config is set to force https, so I'm pretty sure my connections are secure, at least up to the proxy.

Is there something obvious that I'm missing here? [Looking at the Koa source code](https://github.com/koajs/koa/blob/757e4ebc5146784ebd00664545c6d37231610a61/lib/context.js#L171), I'm guessing `request.secure = false` but then why would `ctx.secure = true`?

**Comment by manimovassagh** [comment]
The issue is that `ctx.secure` and `ctx.cookies.secure` are two different things, and Koa's Cookies constructor doesn't automatically inherit the request's secure status when `proxy: true` is set.

Here's what happens:

1. `ctx.secure` checks `ctx.request.secure`, which looks at `ctx.protocol === "https"`. With `proxy: true`, Koa reads the `X-Forwarded-Proto` header, so `ctx.secure` correctly returns `true` behind Fly.io's proxy.

2. However, `ctx.cookies` is a separate `Cookies` instance created with `req.connection.encrypted` (the raw Node.js socket), which is `false` behind a reverse proxy because the TLS is terminated at the proxy.

So `ctx.secure === true` but `ctx.cookies.secure === false` — that's the mismatch.

Your workaround is actually the correct fix. It's not a hack — it's the documented way to handle this. You need to tell the Cookies instance that the connection is secure:

```js
app.use((ctx, next) => {
  ctx.cookies.secure = ctx.secure;
  return next();
});
```

Alternatively, starting with `koa` 2.15+ and `cookies` 0.9+, you can pass `secure: true` in the cookie options directly:

```js
ctx.cookies.set("session", value, {
  secure: ctx.secure,
  sameSite: ctx.secure ? "none" : "strict",
  httpOnly: true,
});
```

As for security: yes, `ctx.secure === true` behind a TLS-terminating proxy means the connection from client to proxy is encrypted, which is the part that matters. The proxy-to-app hop is internal (Fly.io runs it over their private network). This is standard practice — it's how every cloud provider (AWS ALB, GCP Load Balancer, Cloudflare, etc.) handles TLS.

Regarding your last question about the Koa source: `ctx.secure` delegates to `ctx.request.secure` (which checks the protocol including `X-Forwarded-Proto` when `proxy: true`), while `context.js` line 171 is a different getter. They're not the same code path, which explains the discrepancy.
